File: Programmers/codingTest_Kit/Heap/42627.cpp
#include <string>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

int solution(vector<vector<int>> jobs) {
    int answer = 0;
    sort(jobs.begin(), jobs.end());
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> PQ;
    int idx = 0, time = 0;
    int length = jobs.size();
    while(idx < length || !PQ.empty()){
        if(idx < length && time >= jobs[idx][0]){
            PQ.push({jobs[idx][1], jobs[idx][0]});
            idx++;
            continue;
        }
        if(!PQ.empty()){
            time += PQ.top().first;
            answer += time - PQ.top().second;
            PQ.pop();
        }
        else{
            time = jobs[idx][0];
        }
    }

    return answer / length;
}

File: Programmers/codingTest_Kit/Heap/42627.java
import java.util.*;

class Solution {
    public int solution(int[][] jobs) {
        int answer = 0;
        Arrays.sort(jobs, (x, y) -> (x[0] - y[0]));
        PriorityQueue<int[]> pq = new PriorityQueue<int[]>((x, y) -> x[1] - y[1]);
        int N = jobs.length;
        int time = 0, index = 0;
        while (!pq.isEmpty() || index < N) {

            while (index < N && jobs[index][0] <= time) {
                pq.add(jobs[index++]);
            }

            if (pq.isEmpty()) {
                time = jobs[index][0];
                continue;
            }

            int[] job = pq.poll();
            time += job[1];
            answer += time - job[0];

        }

        return answer / N;
    }
}